### PR TITLE
Fixed #21039 -- Introduce AddIndexConcurrently operation for PostgreSQL

### PR DESCRIPTION
**Pull request by dtao**
This adds a `CreateIndexConcurrently` operation to the django.contrib.postgres.operations module. This operation raises a `NotSupportedError` if you attempt to run it as part of an atomic migration. It inherits from `AddIndex` and should effectively be a drop-in replacement as long as a migration has `atomic = False` set.

### CODE REVIEW

File: tests/migrations/test_operations.py [hunk 1]
```diff
@@ -2953,3 +2956,57 @@ def test_reference_field_by_through_fields(self):
         self.assertIs(operation.references_field('Through', 'whatever'), False)
         self.assertIs(operation.references_field('Through', 'first'), True)
         self.assertIs(operation.references_field('Through', 'second'), True)
+
+
+@unittest.skipUnless(connection.vendor == 'postgresql', 'CreateIndexConcurrently is PostgreSQL-specific')
+class CreateIndexConcurrentlyTests(OperationTestBase):
```
Comment (dtao) [inline tests/migrations/test_operations.py]: A note on these tests being in this file: I initially attempted to add a test file in the tests/postgres_tests directory, but I wanted to use some of the functionality provided by `OperationTestBase` and struggled with figuring out a clean way to import the class without causing errors. If this poses a problem, I am certainly willing to try again; I just might need some guidance.

Comment (dtao) [reply tests/migrations/test_operations.py]: Update: I moved these after all. It did require me to move `OperationTestBase` into tests/migrations/test_base.py.

File: tests/migrations/test_base.py [hunk 2]
```diff
@@ -130,3 +132,154 @@ def temporary_migration_module(self, app_label='migrations', module=None):
                 new_module = os.path.basename(target_dir) + '.migrations'
                 with self.settings(MIGRATION_MODULES={app_label: new_module}):
                     yield target_migrations_dir
+
+
+class OperationTestBase(MigrationTestBase):
```
Comment (dtao) [inline tests/migrations/test_base.py]: As explained in [d68bafc](https://github.com/django/django/pull/10901/commits/d68bafca1f8b2e34dcc80ba0beaa02103fed0235), I moved this class here to make it more readily importable elsewhere (namely, in tests/postgres_tests/test_operations.py).

**Comment by dtao** [comment]
Notes:

1. I've [added documentation](https://github.com/django/django/pull/10901/commits/b98a0db0f0d33852a74c0ce53d44ddbd1af74d9b), per the [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist). Let me know if I missed a spot!
2. It seems [no new features are being accepted for 2.2](https://www.djangoproject.com/weblog/2019/feb/11/django-22-beta-1-released/), so I indicated in the documentation that this functionality is added in 3.0. Let me know if there's anything else I need to do to target this pull request properly.
3. I understand that the convention is to squash changes down to a single commit before merging. I have refrained from doing that _yet_ since I find it can make reviewers' lives more difficult in case they've already reviewed some of the code. I figure the appropriate time to squash is after everything's been approved and it's time to merge. Let me know if that isn't the right approach.

**Comment by MarkusH** [comment]
Hi @dtao, thank you for your contribution.

I like that you're tackling that ticket. I have one concern about the specific implementation.

As it is right now, one can only build a b-tree index concurrently, but not a gin/gist/brin/…. From an API perspective, I'd prefer having a `concurrent=True` flag on the `AddIndex` operation instead of having a dedicated `CreateIndexConcurrently` operation.

You could leave the `OperationTestBase` in `tests/migrations/test_operations.py` and skip tests when they're not run on PostgreSQL by decorating them with `@unittest.skipUnless(connection.vendor == 'postgresql', "PostgreSQL specific tests")

**Comment by dtao** [comment]
@MarkusH If you review [the conversation on #21039](https://code.djangoproject.com/ticket/21039#comment:12) you'll see that your suggestion was considered, albeit briefly. My reasoning for preferring a custom class was that, to my knowledge, PostgreSQL is at present the only database backend that supports this functionality; and adding a parameter to the generic base class that's only supported by a single vendor seemed (to me) like adding unnecessary complexity to the general-purpose interface.

That said, this is certainly a good catch:

> As it is right now, one can only build a b-tree index concurrently, but not a gin/gist/brin/….

From what I can tell, solving this issue does not require a fundamentally different approach. Since `CreateIndexConcurrently` already accepts an `index` parameter, it just needs to adapt the SQL produced by calling `create_sql` on that index (which should do the right thing for PostgreSQL-specific indexes like `GinIndex`, `GistIndex`, &c.).

I'm going to try out that approach and see how far I get.

> You could leave the `OperationTestBase` in `tests/migrations/test_operations.py` and skip tests when they're not run on PostgreSQL by decorating them with

I am sure that some variation on this method is possible. Just for the record, though, while I did attempt this initially, I ultimately opted to move `OperationTestBase` instead because:

1. Simply importing any modules under `django.contrib.postgres` leads to errors in environments without psycopg2 installed (so taking this approach at all requires either importing the necessary modules inline or having `except Import Error: pass` somewhere, which strikes me as a warning sign); and
2. There is currently no precedent for vendor-specific tests in the `tests.migrations` module. Obviously this doesn't mean there could never be. But from an organizational perspective, I would prefer to respect the convention that appears to exist today, where there is a `tests.postgres_tests` module for testing functionality specific to PostgreSQL.

**Comment by dtao** [comment]
Update: as @MarkusH helpfully [pointed out](https://github.com/django/django/pull/10901#issuecomment-464780525), the original implementation only supported creating indexes using the default B-tree method. I've updated the code to be compatible with other index methods in PostgreSQL, e.g. `BrinIndex`, `GinIndex`, etc.

As a bonus, the code is now simpler!

**Comment by MarkusH** [comment]
Thanks for the update, @dtao!

Generally looks good to me, I'd need to take a closer look, though. I have two more suggestions, though:

1. The class to create an index concurrently is now called `CreateIndexConcurrently` which seems a bit … weird when its superclass is called `AddIndex`. Did you consider the name `AddIndexConcurrently`?
1. There's currently no way to remove an index concurrently, other than in a reverse `CreateIndexConcurrently` step. How about adding a `DropIndexConcurrently` or `RemoveIndexConcurrently` operation that works like `RemoveIndex`?

**Comment by dtao** [comment]
> which seems a bit … weird when its superclass is called `AddIndex`

Good point! To be honest, this only dawned on me as I was making the last set of changes. I agree that `AddIndexConcurrently` would be a better name. The simple explanation is that `CreateIndexConcurrently` is adapted from some code that was originally written before `AddIndex` existed.

> How about adding a `DropIndexConcurrently` or `RemoveIndexConcurrently` operation

Sure, I can add that. Incidentally, now that I'm exploring this code, I see that there is an interesting asymmetry in how removing an index works compared to adding an index: `sql_delete_index` in PostgreSQL's schema editor class uses `DROP INDEX IF EXISTS` while `sql_create_index` just uses `CREATE INDEX`. (Maybe this is because `DROP INDEX IF EXISTS` has been available [since PostgreSQL 8.2](https://www.postgresql.org/docs/8.2/sql-dropindex.html) whereas `CREATE INDEX IF NOT EXISTS` was only [introduced in 9.5](https://www.postgresql.org/docs/9.5/sql-createindex.html).)

In any case, this means that a migration to add an index will fail if an index with the same name already exists; but a migration to drop an index that _doesn't_ already exist will succeed.

My intuition is that, while this behavior is strange to me, I should preserve it for `RemoveIndexConcurrently`.

**Comment by dtao** [comment]
Update:

1. Renamed `CreateIndexConcurrently` to `AddIndexConcurrently`
2. Added `RemoveIndexConcurrently` operation
3. Updated the documentation

**Comment by MarkusH** [comment]
> > How about adding a `DropIndexConcurrently` or `RemoveIndexConcurrently` operation
> 
> Sure, I can add that. Incidentally, now that I'm exploring this code, I see that there is an interesting asymmetry in how removing an index works compared to adding an index: `sql_delete_index` in PostgreSQL's schema editor class uses `DROP INDEX IF EXISTS` while `sql_create_index` just uses `CREATE INDEX`. (Maybe this is because `DROP INDEX IF EXISTS` has been available [since PostgreSQL 8.2](https://www.postgresql.org/docs/8.2/sql-dropindex.html) whereas `CREATE INDEX IF NOT EXISTS` was only [introduced in 9.5](https://www.postgresql.org/docs/9.5/sql-createindex.html).)
> 
> In any case, this means that a migration to add an index will fail if an index with the same name already exists; but a migration to drop an index that _doesn't_ already exist will succeed.
> 
> My intuition is that, while this behavior is strange to me, I should preserve it for `RemoveIndexConcurrently`.

IIRC, this was by design when we introduced indexes. But I don't find any written records of that anywhere. But yeah, since that'd be a backward incompatible change, please preserve that behavior.

**Comment by dtao** [comment]
I [pushed a fix](https://github.com/django/django/pull/10901/commits/ff7e8f70c4db79ca8f6d19c64af349c5aeaeabfe) yesterday for [the failing docs build](https://djangoci.com/job/docs/18821/); but it seems that didn't trigger another test run.

Is there anything I need to do to re-run the docs check?

**Comment by auvipy** [comment]
the doc failes seems un related

**Comment by timgraham** [comment]
Please structure as two commits: the first should move `OperationTestBase`, the second is the new feature.

File: tests/postgres_tests/test_operations.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        app_label = 'test_pgdic'
+
+        project_state = self.set_up_test_model(app_label, index=True)
+        new_state = project_state.clone()
+
+        operation = RemoveIndexConcurrently('Pony', 'pony_pink_idx')
+        operation.state_forwards(app_label, new_state)
+        self.assertEqual(len(new_state.models[app_label, 'pony'].options['indexes']), 0)
+
+        definition = operation.deconstruct()
```
Comment (timgraham) [inline tests/postgres_tests/test_operations.py]: `name, args, kwargs = operation.deconstruct()`

File: tests/postgres_tests/test_operations.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        with connection.schema_editor(atomic=False) as editor:
+            operation.database_backwards(app_label, editor, new_state, project_state)
+
+        self.assertIndexNotExists('%s_pony' % app_label, ['pink'])
+
+
+@unittest.skipUnless(connection.vendor == 'postgresql', 'RemoveIndexConcurrently is PostgreSQL-specific')
+@modify_settings(INSTALLED_APPS={'append': 'migrations'})
+class RemoveIndexConcurrentlyTests(OperationTestBase):
+    def test_remove_index_concurrently_description(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_operations.py]: Remove `_remove_index_concurrently` from test names as it's redundant with the class name.

File: tests/postgres_tests/test_operations.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    from django.contrib.postgres.operations import AddIndexConcurrently, RemoveIndexConcurrently
+    from django.contrib.postgres.indexes import BrinIndex
+except ImportError:
+    pass
+
+
+@unittest.skipUnless(connection.vendor == 'postgresql', 'AddIndexConcurrently is PostgreSQL-specific')
+@modify_settings(INSTALLED_APPS={'append': 'migrations'})
+class AddIndexConcurrentlyTests(OperationTestBase):
+    def test_add_index_concurrently_requires_name(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_operations.py]: I don't think the behavior inherited from the superclass needs to be tested again.

File: docs/ref/contrib/postgres/operations.txt [hunk 6]
```diff
@@ -98,3 +98,33 @@ run the query ``CREATE EXTENSION IF NOT EXISTS hstore;``.
...
+.. class:: AddIndexConcurrently(model_name, index)
+
+    .. versionadded:: 3.0
+
+    Like :class:`~django.db.migrations.operations.AddIndex`, creates an index
+    in the database table for the model with ``model_name``, but with the
+    ``CONCURRENTLY`` option in PostgreSQL. ``index`` is an instance of the
+    :class:`~django.db.models.Index` class.
+
+    Note that the ``CONCURRENTLY`` option is not supported inside a
```
Comment (timgraham) [inline docs/ref/contrib/postgres/operations.txt]: Chop "Note that"

File: django/contrib/postgres/operations.py [hunk 7]
```diff
@@ -75,3 +79,63 @@ class UnaccentExtension(CreateExtension):
...
+        schema_editor.execute(sql)
+
+    def database_backwards(self, app_label, schema_editor, from_state, to_state):
+        self._ensure_not_in_transaction(schema_editor)
+
+        model = from_state.apps.get_model(app_label, self.model_name)
+        sql = re.sub('^DROP INDEX', 'DROP INDEX CONCURRENTLY', str(self.index.remove_sql(model, schema_editor)))
+        schema_editor.execute(sql)
+
+    def _ensure_not_in_transaction(self, schema_editor):
```
Comment (timgraham) [inline django/contrib/postgres/operations.py]: It seems a mixin would be useful to avoid duplicating shared behavior such as this method.

**Comment by felixxm** [comment]
Updated in #11594.

**Pull request closed without merging**